lib/loggability/log_device/datadog.rb

```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env ruby
+# vim: set nosta noet ts=4 sw=4:
+# encoding: utf-8
+
+require 'socket'
+require 'net/https'
+require 'json'
+require 'concurrent'
+require 'loggability/logger' unless defined?( Loggability::Logger )
+
+require 'loggability/http_log_device'
+
+# A log device that sends logs to Datadog's HTTP endpoint
+# for receiving logs
+class Loggability::LogDevice::Datadog < Loggability::HttpLogDevice
+
+	### Datadog's HTTP endpoint URL for sending logs to
+	ENDPOINT = "https://http-intake.logs.datadoghq.com"
+
+	### Datadogs API endpoint path for receiving logs
+	PATH = "/v1/input/"
+
+	### The max number of messages that can be sent to datadog in a singole payload
+	MAX_BATCH_SIZE = 500
+
+	### The max size in bytes for a single message.
+	### Limiting the message size to 200kB  to leave room for other info such as tags, metadata, etc.
+	### DataDog's max size for a single log entry is 256kB
+	MAX_MESSAGE_BYTESIZE = 204_800
+
+	### The max size in bytes of all messages in the batch.
+	### Limiting the total messages size to 4MB to leave room for other info such as tags, metadata, etc.
+	### Datadog's max size for the entire payload is 5MB
+	MAX_BATCH_BYTESIZE = 4_194_304
+
+
+	def initialize( api_key )
+		super( ENDPOINT )
+		self.target = api_key
+		self.setup_logs_queue
+		self.batched_messages_bytesize = 0
+	end
+
+
+	######
+	public
+	######
+
+	## the thread-safe array for batching log messages
+	attr_reader :logs_queue
+
+	# The target of the log device
+	attr_reader :target
+
+	## This the size in bytes of batched messages. It cannot exceed 5MB, Datadogs limit per payload
+	attr_reader :batched_messages_bytesize
+
+
+	### Sets up the data for the log message and sends it to datadog
+	### From: https://docs.datadoghq.com/api/?lang=bash#send-logs-over-http
+	###     Maximum content size per payload: 5MB
+	###     Maximum size for a single log: 256kB
+	###     Maximum array size if sending multiple logs in an array: 500 entries
+	def write( msg )
+		raise "Log message size cannot exceed %d bytes" % [ MAX_MESSAGE_BYTESIZE ] if msg.bytesize > MAX_MESSAGE_BYTESIZE
+
+		self.batched_messages_bytesize += msg.bytesize
+
+		self.logs_queue.push({
+			hostname: Socket.gethostname,
+			message: msg
+		})
+
+		## send the batch of logs if it is approaching datadog's payload size limits
+		if self.logs_queue.size == MAX_BATCH_SIZE || self.batched_messages_bytesize > MAX_BATCH_BYTESIZE
+			self.send_logs
+			self.batched_messages_bytesize = 0
+		end
+	end
+
+	### Closes the http connection
+	def close
+		begin
+			self.http.finish
+		rescue IOError
+			## ignore it since http session has not yet started.
+		end
+	end
+
+
+	##########
+	protected
+	##########
+
+
+	def batched_messages_bytesize=( size )
+		@batched_messages_bytesize = size
+	end
+
+	###
+	### sends all a batch of up to  the batched log messages asynchronously to datadog
+	def send_logs
+		return if self.logs_queue.size == 0
+
+		batched_messages = Concurrent::Array.new
```

Comment:
I don't think we need this array to be thread-safe, right? It's going to be allocated, filled and read all within a single thread with a scope local to this method. Sometimes I get my wired crossed with this stuff though.